lib/ast-nested-transform.js

```diff
@@ -0,0 +1,88 @@
+'use strict';
+
+const reLines = /(.*?(?:\r\n?|\n|$))/gm;
+const ALPHA = /[A-Za-z]/;
+
+class AngleBracketPolyfill {
+  constructor(options) {
+    this.syntax = null;
+    this.sourceLines = options.contents && options.contents.match(reLines);
+  }
+
+  transform(ast) {
+    let b = this.syntax.builders;
+
+    function dasherize(string) {
+      return string.replace(/[A-Z]/g, function(char, index) {
+        if (index === 0 || !ALPHA.test(string[index - 1])) {
+          return char.toLowerCase();
+        }
+
+        return `-${char.toLowerCase()}`;
+      });
+    }
+
+    function replaceNestedComponents(string) {
+      return string.replace('::', '/');
+    }
+
+    function getBlockParamName(node) {
+      let unnestedName = node.tag.replace('::', '');
+      let possibleName = unnestedName;
+
+      const nestedNames = node.children
```

Comment:
I don't think this guard is quite correct, there are a few cases not accounted for:

* the given name _could_ be used by any number of nested children below us (this only checks direct children)
* this is only checking for invocations using the possible name, but not for things like path expressions, yielded block params, etc
* the name we pick could be clobbering a name that is in use in the outer scope (e.g. above us)
* the logic doesn't currently support `<Foo::Bar::Baz />` (`string.replace('::', '/')` only replaces the first occurrence of `::`)

Also, the logic to add the `let` will actually add many more `let` blocks than we might want. Consider:

```hbs
<Foo::Bar />
<Foo::Bar />
<Foo::Bar />
<Foo::Bar />
<Foo::Bar />
<Foo::Bar />
```

With the current transform, ^ template would end up with many "extra" `let` blocks.

Reply:
I believe the most recent round of updates that I just pushed address all of these concerns...